### DISCUSSION TITLE
Whether all platforms are supported is deterministic

### DISCUSSION THREAD

**Discussion by PinkFatMan** [jrouwe/JoltPhysics#617] [Q&A] [answered]
Hi, I am making a Moba game using unity. A deterministic physics engine needs to be integrated to achieve the same simulation results with the same input across all platforms. Is Jolt a deterministic physics engine?

**Comment by jrouwe** [comment]
Yes it is, please go to https://jrouwe.github.io/JoltPhysics/ to the Deterministic Simulation section.

**Comment by PinkFatMan** [comment, marked as answer]
Is it okay to modify the CMakeLists.txt file? You still need to copy the option (CROSS-PLATFORM-DETERMINISTIC "Cross platform deterministic" OFF) in CMakeLists.txt to the Jolt.cmake file. Sorry, I'm not very familiar with make

**Reply by jrouwe** [reply]
Just add -DCROSS_PLATFORM_DETERMINISTIC=ON to the command line when you generate a make file, e.g. under Linux you can do:

```
./cmake_linux_clang_gcc.sh Debug clang++ -DCROSS_PLATFORM_DETERMINISTIC=ON
```

under Windows:

```
cmake_vs2022_cl.bat -DCROSS_PLATFORM_DETERMINISTIC=ON
```

**Reply by PinkFatMan** [reply]
Hello, from https://jrouwe.github.io/JoltPhysics/ The Deterministic Simulation section of the document explains Architecture (x86 or ARM) Note that 32 bit architectures are currently not compatible with 64 bit architectures. If I want to ensure certainty across all platforms. Do you need to use fixed-point math to replace all floating-point numbers?

**Reply by jrouwe** [reply]
In general you don't need to use fixed-point math to get determinism across platforms. It does require careful programming.

What is 'all platforms' in your case?

**Reply by PinkFatMan** [reply]
What I mean by all platforms is that I can simulate the same input on windows,mac,ios, Android,linux and get the same results. As we often play competitive games, they can all support players with different platforms in the same game. However, the simulation results are guaranteed to be the same for all players

**Reply by PinkFatMan** [reply]
But this has nothing to do with the platform structure. Because if the architectures are not compatible then can players of both architectures not play together?

**Reply by jrouwe** [reply]
I would assume that you are going to produce a 64-bit executable on Windows, macOS, iOS, Android and Linux. If this is the case then the simulation will be the same on all platforms.

**Reply by PinkFatMan** [reply]
Yes, my requirement should be that generating 32-bit and 64-bit on the windows platform also guarantees certainty. What do I need to do?

**Reply by jrouwe** [reply]
#define JPH_ENABLE_DETERMINISM_LOG, run ```PerformanceTest -q=LinearCast -t=1 -s=ConvexVsMesh``` and start comparing the logs between 32 and 64 bit mode to see where the bits start to deviate, add more JPH_DET_LOG statements until you narrow down where the difference comes from. It might be a very simple thing, I haven't spent any effort on this.

**Reply by jrouwe** [reply]
Actually it looks like it was trivial: https://github.com/jrouwe/JoltPhysics/pull/621

**Reply by PinkFatMan** [reply]
Thanks for your reply, it has been confirmed to be available after log comparison. And subsequent non-synchronization can be found in a similar way.
